pkg/sfu/buffer/rtpstats_receiver.go

```diff
@@ -468,4 +473,62 @@ func (r *RTPStatsReceiver) ToProto() *livekit.RTPStats {
 	)
 }
 
+func (r *RTPStatsReceiver) getOutOfOrderHistorySlot(esn uint64, ehsn uint64) (int, int) {
+	diff := int64(ehsn - esn)
+	if diff >= cHistorySize || diff < 0 {
+		// too old OR too new (i. e. ahead of highest)
+		return -1, -1
+	}
+
+	return int(esn) % len(r.history), int(esn & 63)
+}
+
+func (r *RTPStatsReceiver) getHistorySlot(esn uint64, ehsn uint64) (int, int) {
+	if int64(esn-ehsn) < 0 {
+		return r.getOutOfOrderHistorySlot(esn, ehsn)
+	}
+
+	return int(esn) % len(r.history), int(esn & 63)
+}
+
+func (r *RTPStatsReceiver) setHistory(esn uint64, ehsn uint64) {
+	slot, offset := r.getHistorySlot(esn, ehsn)
+	if slot < 0 {
+		return
+	}
+
+	r.history[slot] |= (1 << offset)
+}
+
+func (r *RTPStatsReceiver) clearHistory(extStartInclusive uint64, extEndExclusive uint64, ehsn uint64) {
+	if extEndExclusive <= extStartInclusive {
+		return
+	}
+
+	slot, offset := r.getHistorySlot(extStartInclusive, ehsn)
+	if slot < 0 {
+		return
+	}
+	for esn := extStartInclusive; esn != extEndExclusive; esn++ {
```

Comment:
if this has to handle ranges larger than 64 it might be faster to operate on words rather than bits. something like this https://go.dev/play/p/veal-ypouwO maybe? we should probably revisit your suggestion to consolidate our bitmap implementations

Reply:
Yeah, good point. I will pick this up in part 2. Yeah, we should make a util. Will have a look at that as I make progress on this.